calculators/_memoria_matchups.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+$GLOBALS['rating_calculators']['_memoria_matchups'] = [
+    'description' => '
+        by <a href="/player/125980">Memoria</a>: per-matchup calculation v1
+        
+        - only 1v1, 2v2, and 3v3
+        - games under 4 minutes don\'t count
+        - players start at 2000 rating, 4000 rating difference = 90% win prediction, max 250 rating per match
+        - win prediction is calculated for each matchup (1v1), then it\'s combined into a final prediction 
+        ',
+    'default_rating' => 2000,
+    'fields' => [
+        'g.duration',
+    ],
+    'main' => '_memoria_matchups__main',
+    ];
+
+function _memoria_matchups__main(&$teams) {
+    if (count($teams[0]) !== count($teams[1])) {
+        return;
+    }
+    
+    $game_duration_seconds = floor($teams[0][0]['duration'] / 22.4);
+    if ($game_duration_seconds < 240) {
+        return;
+    }
+    
+    $factor = 4000;
+    $matchup_predictions = [];
+    $matchup_predictions_weight = [];
+    $matchup_predictions_weight_total = 0;
+    for ($i = 0; $i < count($teams[0]); $i++) {
+        $elo_part = abs(($teams[0][$i]['rating'] - $teams[1][$i]['rating']) / $factor);
+        $matchup_predictions[$i] = floor(1000 / (1 + 10 ** ($elo_part))) / 1000;
+		
+        if ($teams[0][$i]['rating'] > $teams[1][$i]['rating']) {
+            $matchup_predictions[$i] = 1 - $matchup_predictions[$i];
+        }
+        
+        $matchup_predictions_weight[$i] = ((abs(0.5 - $matchup_predictions[$i]) + 0.015) * 6) ** ($elo_part * 1.5);
+        $matchup_predictions_weight_total += $matchup_predictions_weight[$i];
+    }
+    
+    $prediction = 0;
+    for ($i = 0; $i < count($teams[0]); $i++) {
+        $normalised_weight_mult = $matchup_predictions_weight[$i] / $matchup_predictions_weight_total;
```

Comment:
you start with $matchup_predictions_weight_total=0, if something goes wrong, and a team will have 0 players, the code will break, instead of simply not assigning any rating.

then, even if teams are not empty, how do you guarantee `((abs(0.5 - $matchup_predictions[$i]) + 0.015) * 6) ** (abs(($teams[0][$i]['rating'] - $teams[1][$i]['rating']) / $factor) * 1.5)`? Right now it can be followed through formulas, and maybe your if() guarantees it (I've lost it there), but this is way too complicated to be sustainable. Even if it ends up guaranteeing it now, next time you try to improve algorithm and break this brittle balance.

Make absolutely sure it is never 0. Code must not break. Even if($total === 0) {$total = 1;} would work, although it's a hack to an algorithm.

Reply:
This is just not how you should write your code. To use a divisor, that is 0, and then stars must align to keep changing it to nonzero, over chained formulas, without any precautions like `if(divisor==0) then return;`

It is also not sustainable, because right now you, it seems, indeed achieved always non-zero divisor. But a year later, you, or even worse someone else will want do adjust this code, "this +0.015 coefficient is throwing balance off, I think I'll remove it", but this was the only thing that kept the code from exploding, who would ever know?

Since you do not agree, sure, I'll accept the way it is, first time your code really kills the site, it will be easier for me to argue :)